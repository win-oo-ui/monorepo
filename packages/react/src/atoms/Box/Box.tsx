import { Spacing, PositionType } from "@winoo/foundation";
import React, { ReactNode, useMemo } from "react";

export interface HelperPositionProps extends PositionType {
  value?: keyof typeof Spacing;
}

interface BoxProps {
  children: ReactNode;
  margin?: HelperPositionProps;
  padding?: HelperPositionProps;
  elementType?: keyof JSX.IntrinsicElements;
}

const Box: React.FC<BoxProps> = ({
  children,
  margin,
  padding,
  elementType: Wrapper = "div",
}) => {
  const classes = useMemo(() => {
    let modClasses = ``;

    if (
      margin?.value &&
      !(margin?.top || margin?.bottom || margin?.left || margin?.right)
    ) {
      modClasses += ` wo-margin-${margin.value}`;
    }

    if (margin?.left) {
      modClasses += ` wo-margin-left-${margin.left}`;
    }

    if (margin?.right) {
      modClasses += ` wo-margin-right-${margin.right}`;
    }

    if (margin?.top) {
      modClasses += ` wo-margin-top-${margin.top}`;
    }

    if (margin?.bottom) {
      modClasses += ` wo-margin-bottom-${margin.bottom}`;
    }

    if (padding?.value) {
      modClasses += ` wo-padding-${padding.value}`;
    }

    if (padding?.left) {
      modClasses += ` wo-padding-left-${padding.left}`;
    }

    if (padding?.right) {
      modClasses += ` wo-padding-right-${padding.right}`;
    }

    if (padding?.top) {
      modClasses += ` wo-padding-top-${padding.top}`;
    }

    if (padding?.bottom) {
      modClasses += ` wo-padding-bottom-${padding.bottom}`;
    }

    return modClasses;
  }, [margin]);

  return <Wrapper className={classes}>{children}</Wrapper>;
};

export default Box;
