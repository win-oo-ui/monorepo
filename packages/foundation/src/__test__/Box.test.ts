import Box from "../Box";

test("Snapshop of Boxs: Foundation", () => {
  expect(Box).toMatchSnapshot();
});
