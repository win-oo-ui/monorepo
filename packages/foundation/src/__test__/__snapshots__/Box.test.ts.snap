// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshop of Boxs: Foundation 1`] = `
{
  "div": "div",
  "h1": "h1",
  "h2": "h2",
  "h3": "h3",
  "h4": "h4",
  "h5": "h5",
  "h6": "h6",
  "p": "p",
  "span": "span",
}
`;
